### PR TITLE
Use packed data within ChineseBasedYearInfo

### PR DESCRIPTION
**Pull request by Manishearth**
Part of https://github.com/unicode-org/icu4x/issues/3933

This does not yet add data loading, but it gets very close to it.

This PR makes the size of Date not bloat too much, without requiring much computation for any of the getters -- they're all basic bit operations.

### CODE REVIEW

File: components/calendar/src/chinese_based.rs [hunk 1]
```diff
@@ -93,97 +112,126 @@ impl<CB: ChineseBased> PrecomputedDataSource<ChineseBasedYearInfo>
...
+    fn last_day_of_month(self, month: u8) -> u16 {
+        let months = u16::from_le_bytes([self.0, self.1]);
+        // month is 1-indexed, so `29 * month` includes the current month
+        let mut prev_month_lengths = 29 * month as u16;
+        // month is 1-indexed, so `1 << month` is a mask with all zeroes except
+        // for a 1 at the bit index at the next month. Subtracting 1 from it gets us
+        // a bitmask for all months up to now
+        let long_month_bits = months & ((1 << month as u16) - 1);
+        prev_month_lengths += long_month_bits.count_ones().try_into().unwrap_or(0);
+        prev_month_lengths
```
Comment (sffc) [inline components/calendar/src/chinese_based.rs:193]: Praise: clever use of `count_ones()`

**Review by sffc — APPROVED** [review]
Nice

File: components/calendar/src/chinese_based.rs [hunk 2]
```diff
@@ -93,97 +112,126 @@ impl<CB: ChineseBased> PrecomputedDataSource<ChineseBasedYearInfo>
...
+/// Byte 0:          [  month lengths .............
+/// Byte 1:         .. month lengths ] | [ leap month index ..
+/// Byte 2:          ] | [   NY offset   ] | unused
 /// ```
 ///
 /// Where the New Year Offset is the offset from ISO Jan 21 of that year for Chinese New Year,
 /// the month lengths are stored as 1 = 30, 0 = 29 for each month including the leap month.
-#[derive(Debug, Copy, Clone)]
+///
+/// Should not
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/calendar/src/chinese_based.rs:127]: ... I have absolutely no idea

File: components/calendar/src/chinese_based.rs [hunk 3]
```diff
@@ -93,97 +112,126 @@ impl<CB: ChineseBased> PrecomputedDataSource<ChineseBasedYearInfo>
...
-            if is_long {
-                30
-            } else {
-                29
+    fn new(month_lengths: [bool; 13], leap_month_idx: Option<NonZeroU8>, ny_offset: u8) -> Self {
+        debug_assert!(
+            !month_lengths[12] || leap_month_idx.is_some(),
+            "Last month length should not be set for non-leap years"
+        );
+        debug_assert!(ny_offset < 32, "Year offset too big to store");
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/calendar/src/chinese_based.rs:137]: Potentially 31, but yes. Andrew and I spent some time researching this but I do plan to run a test on every year to check this property.

File: components/calendar/src/chinese_based.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/calendar/src/chinese_based.rs:127]: I'll fix this in the upcoming PR.

**Pull request merged**